feature(rome_js_formatter): Add support for printing single quotes

cpojer opened the pull request:
## Summary
This PR adds a `--quote-style` option to the Rome Formatter so that users can choose between single and double quotes. Arguably, template literals should also be an available option, but that can be discussed later.

## Test Plan

* `cargo test`
* `cargo run -p rome_cli -- format --quote-style=single <file>` works.

cpojer commented on crates/rome_js_formatter/src/utils/mod.rs:
```diff
@@ -614,16 +614,31 @@ pub(crate) fn format_string_literal_token(
-        match normalize_newlines(&s, ['\r']) {
-            Cow::Borrowed(_) => s,
-            Cow::Owned(s) => s,
+    let content = match formatter.options().quote_style {
+        QuoteStyle::Double => {
+            if quoted.starts_with('\'') && !quoted.contains('"') {
+                let s = &quoted[1..quoted.len() - 1];
+                let s = format!("\"{}\"", s);
+                match normalize_newlines(&s, ['\r']) {
+                    Cow::Borrowed(_) => s,
+                    Cow::Owned(s) => s,
+                }
+            } else {
+                normalize_newlines(quoted, ['\r']).into_owned()
+            }
+        }
+        QuoteStyle::Single => {
+            if quoted.starts_with('"') && !quoted.contains('\'') {
+                let s = &quoted[1..quoted.len() - 1];
+                let s = format!("'{}'", s);
+                match normalize_newlines(&s, ['\r']) {
+                    Cow::Borrowed(_) => s,
+                    Cow::Owned(s) => s,
+                }
+            } else {
+                normalize_newlines(quoted, ['\r']).into_owned()
+            }
```
How do I make this look nice?

kassens replied:
👋 was just taking a look at Rome and surprise do see you around.

start with this, then use the them in the code to avoid the duplication.
```
let (primary_quote_char, secondary_quote_char) = match quote_style {
  Style::Double => ('"', '\''),
  Style::Single => ('\'', '"'),
};
```

cpojer replied:
Nice to see you here ❤️ Thanks, nice destructuring syntax. I cleaned it up.

cpojer commented on crates/rome_js_formatter/tests/specs/js/module/string/string.js.snap:
```diff
@@ -0,0 +1,167 @@
+import foo from 'foo.json' assert { type: 'json' };
+import foo from 'foo.json' assert { type: 'json' };
+import foo2 from 'foo.json' assert {
+	'type': 'json',
+	type: "html",
+	"type": "js",
```
Any idea why only the first two strings are reprinted with single-quotes in asserts?

ematipico replied:
Because this code is not valid. https://play.rome.tools/?lineWidth=80&indentStyle=tab&indentWidth=2&typescript=false&jsx=false&sourceType=module#aQBtAHAAbwByAHQAIABmAG8AbwAyACAAZgByAG8AbQAgACcAZgBvAG8ALgBqAHMAbwBuACcAIABhAHMAcwBlAHIAdAAgAHsACgAJACcAdAB5AHAAZQAnADoAIAAnAGoAcwBvAG4AJwAsAAoACQB0AHkAcABlADoAIAAiAGgAdABtAGwAIgAsAAoACQAiAHQAeQBwAGUAIgA6ACAAIgBqAHMAIgAsAAoACQB9AA==

Check the errors tab. Broken code won't be formatted

cpojer replied:
Ohhh I see. Well, glad to know my code works but also time for me to learn the ES assert syntax ;)

cpojer commented on crates/rome_js_formatter/tests/specs/js/module/array/array_nested.js.snap:
```diff
@@ -1,8 +1,6 @@
 ---
 source: crates/rome_js_formatter/tests/spec_test.rs
-assertion_line: 157
 expression: array_nested.js
-
```
I don't know what I did to trigger this removal across all files.

Boshen replied:
You have the latest insta, which removed this line, no worries. https://github.com/mitsuhiko/insta/commit/5ab9f4093f6d62d8889918a8a84c99b386f1c816

ematipico commented on website/playground/src/lib.rs:
```diff
@@ -111,6 +113,7 @@ pub fn run(
     let options = FormatOptions {
         indent_style,
         line_width: line_width.try_into().unwrap_or_default(),
+        quote_style: QuoteStyle::from_str(&quote_style).unwrap_or_default(),
```
Here we can use use `quote_style.try_into()` like we do on line 115

cpojer replied:
I didn't implement TryInto for QuoteStyle though, mainly because indentStyle doesn't have it either and I thought this is probably fine. It's only implemented for the u16 → LineWidth conversion.

Let me know if you'd like me to add it, though.

leops replied:
Since string-to-type conversion is impemented with the `FromStr` trait instead of the generic `TryFrom`, it can be called as a method using `parse` instead of `try_into`: `quote_style.parse().unwrap_or_default()`

cpojer replied:
Ah, yay `parse` works!

ematipico commented on crates/rome_cli/src/commands/format.rs:
```diff
@@ -58,6 +58,24 @@ pub(crate) fn format(mut session: CliSession) -> Result<(), Termination> {
 
+    let quote_style = session
+        .args
+        .opt_value_from_str("--quote-style")
```
The help message of the command should be updated with the new option

cpojer replied:
Just added the comment. Would be awesome to auto-generate the help text somehow so this will always be kept in sync.

ematipico reviewed (changes requested):
That's a great contribution! I would be great to add new test cases to make sure we propertly test the "single quote" case. Please follow this guide to how to add new tests: https://github.com/rome/tools/blob/main/crates/rome_js_formatter/docs/write_tests.md

cpojer commented:
@ematipico thanks for the review! I updated the help text.

Re tests, I added one test (which you commented on) where I copy-pasted some other examples together. Please let me know which other tests you'd like me to add.

MichaReiser commented on crates/rome_lsp/src/handlers/formatting.rs:
```diff
@@ -44,6 +44,16 @@ pub(crate) fn to_format_options(
         );
     }
 
+    let custom_ident_style =
```
`custom_quote_style`?

cpojer replied:
Oops, fixing.

MichaReiser commented:
> @ematipico thanks for the review! I updated the help text.
> 
> Re tests, I added one test (which you commented on) where I copy-pasted some other examples together. Please let me know which other tests you'd like me to add.
> 
> If you have instructions on how to try out the cli and VSCode extension in development for testing I can test those flows too!

Thanks @cpojer for this amazing contribution! 

You can run the CLI with `cargo run -p rome_cli -- <args>` I defer to @ematipico for how to test the VS code extension.

leops commented on crates/rome_cli/src/commands/format.rs:
```diff
@@ -58,6 +58,24 @@ pub(crate) fn format(mut session: CliSession) -> Result<(), Termination> {
+            source,
+        })?;
+
+    match quote_style {
```
The `indent_style` had to use a match expression in order to inject the number of spaces into the option, but for this one `if let Some(quote_style) = quote_style { options.quote_style = quote_style; }` should be enough

ematipico reviewed (commented):
At the moment we don't have tests for VSCode, this is something we still need to set up. For now I think it's fine. 

We need to add this option to the [`package.json`](https://github.com/rome/tools/blob/main/editors/vscode/package.json#L84-L102). The name of the option has to be same that has been added to `rome_lsp/src/config.rs`, in camel case, and it should be an enum like `indentStyle`

cpojer commented on crates/rome_js_formatter/tests/specs/js/module/string/string.js:
```diff
@@ -0,0 +1,45 @@
+
+    type:
+        "json" };
+import foo2 from "foo.json" assert { "type": "json", type: "html", "type": "js" };
```
I actually think this is a good test to have though because it will ensure that we don't break this invariant of not formatting strings in broken code.

cpojer commented on crates/rome_lsp/src/handlers/formatting.rs:
```diff
@@ -44,6 +44,16 @@ pub(crate) fn to_format_options(
     }
 
+    let custom_ident_style =
+        QuoteStyle::from_str(workspace_settings.quote_style.as_str()).unwrap_or(QuoteStyle::Double);
```
Ah good point. I changed that for ident style above as well.

cpojer commented:
Thanks for all the thorough feedback. I must admit, I haven't written a single line of Rust in more than two years and as you can clearly see I just started writing code on a Saturday night to solve my own problems so I can start adopting the formatter. Really appreciate all the help here, next up is an option for `bracketSpacing`, after which I can finally start using the Rome formatter :)

ematipico reviewed (approved):
I think we're good with this option. We will take care of the documentation of the website

cpojer commented:
Added the option to the playground, see screenshot:
<img width="1190" alt="Screen Shot 2022-04-11 at 23 23 48" src="https://user-images.githubusercontent.com/13352/162760710-d9c17464-c183-46c0-aa4b-7623fe6f94f2.png">

cpojer commented on website/playground/src/App.tsx:
```diff
@@ -201,16 +220,25 @@ function App() {
+											language: isTypeScript ? 'ts' : 'js',
+											quoteStyle,
 										})}
+										key={
+											code +
+											lineWidth +
+											indentStyle +
+											indentWidth +
+											language +
+											quoteStyle
+										}
```
`formatWithPrettier` is called on every render, but the `<CodeEditor>` component does not re-render when options change. You can see this by changing one of the dropdowns and you'll see that the prettier window does not change until the code is changed again. Here I'm just concatenating all the variables to make it look like a poor-mans's `useEffect`.

MichaReiser replied:
Nit: Can we exclude `code` because the editor is already updating when the `code` change. This would avoid one full unmount & mount (may be noticeable for large files).
